datanode/gateway/middleware.go

```diff
@@ -64,3 +65,55 @@ func MetricCollectionMiddleware(log *logging.Logger, next http.Handler) http.Han
 		metrics.APIRequestAndTimeREST(uri, timetaken.Seconds())
 	})
 }
+
+// Chain builds the middleware Chain recursively, functions are first class
+func Chain(f http.Handler, m ...func(http.Handler) http.Handler) http.Handler {
+	// if our Chain is done, use the original handler func
+	if len(m) == 0 {
+		return f
+	}
+	// otherwise nest the handler funcs
+	return m[0](Chain(f, m[1:cap(m)]...))
+}
+
+func WithAddHeadersMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		iw := &InjectableResponseWriter{ResponseWriter: w}
+		ctx = context.WithValue(ctx, injectableWriterKey, iw)
+		next.ServeHTTP(iw, r.WithContext(ctx))
+	})
+}
+
+type InjectableResponseWriter struct {
+	http.ResponseWriter
+	headers http.Header
+}
+
+type key string
+
+const injectableWriterKey key = "injectable-writer-key"
```

Comment:
We usually inject key like in the following file in the context: libs/context/context.go.

I also kinda seems to remember its the recommended way to do it.